DAOS ADIO driver

mchaarawi opened the pull request.

roblatham00 commented:
You extend ROMIO's general ADIO_End() for hash clieanup.  I'd rather see you do what we did in PVFS2 and create an additional DAOS keyval, and then write delete routines for that keyval. see what we did in src/mpi/romio/adio/ad_pvfs2/ad_pvfs2_common.c

mchaarawi commented:
> You extend ROMIO's general ADIO_End() for hash clieanup. I'd rather see you do what we did in PVFS2 and create an additional DAOS keyval, and then write delete routines for that keyval. see what we did in src/mpi/romio/adio/ad_pvfs2/ad_pvfs2_common.c

Yes thanks for catching that. I had a ticket to move those from the generic common code but that fell through the cracks. fixed now.

roblatham00 commented:
I can no longer rebase this onto master -- I think we snuck a new file system into ROMIO, or maybe it is the change we added to the "try these file systems".  If you could update your changes to the common code, that would be great.   The next step is I think I will just squash all these commits into one big "Initial DAOS driver" commit.   We don't need the merge commits and the 54 other commits are interesting historically but are not relevant to MPICH.    Lastly, we'll run this through the whitespace checker (maint/code-cleanup.sh).    I'm happy to do the squashing and whitespace cleanup.    Does this plan mess up your ongoing development?

mchaarawi commented:
> I can no longer rebase this onto master -- I think we snuck a new file system into ROMIO, or maybe it is the change we added to the "try these file systems". If you could update your changes to the common code, that would be great. The next step is I think I will just squash all these commits into one big "Initial DAOS driver" commit. We don't need the merge commits and the 54 other commits are interesting historically but are not relevant to MPICH. Lastly, we'll run this through the whitespace checker (maint/code-cleanup.sh). I'm happy to do the squashing and whitespace cleanup. Does this plan mess up your ongoing development?

nope, I can rebase with latest master and do some testing today and update the PR.
i'll try and see where the whitespace checker complains. i wasn't aware the script was in the source code (thanks for pointing me to it), i just couldn't access the page from the link above.

mchaarawi commented:
i updated to latest master (wasn't that far behind). i didn't see any conflicts when merging master into my branch and updating the PR.. so not sure why you couldn't do it. Those changes that you mentioned were already in..
Running the code-cleanup.sh  script didn't actually change anything in the files that i added (i rank it recursively from the top of the romio dir).

roblatham00 commented on src/mpi/romio/adio/include/adioi_fs_proto.h:
```diff
@@ -49,4 +50,66 @@ extern struct ADIOI_Fns_struct ADIO_GPFS_operations;
+#ifdef ROMIO_DAOS
+/* prototypes are in adio/ad_daos/ad_daos.h */
+extern struct ADIOI_Fns_struct ADIO_DAOS_operations;
+
```
that's it.  that's all the DAOS bits that should be in adioi_fs_proto.h.   Can 'handle_share' and the enum go in to ad_daos_common.h?  ah, or are these routines piercing the ADIO abstraction layer?

mchaarawi replied:
hmm no, i should be able to move that.. i think i had it here before when i was doing the daos_init / finalize above the adio layer.. will fix shortly.

mchaarawi replied:
so actually i had fixed that before, but im not sure why is it no showing in this PR.
https://github.com/daos-stack/mpich/blob/daos_adio/src/mpi/romio/adio/include/adioi_fs_proto.h

hmm maybe i should just recreate the PR

mchaarawi replied:
are you sure you are not looking at an older version.. looking through the PR diff, i see that was already moved to ad_daos/ad_daos.h

mchaarawi commented:
i had accidentally merged our CI branch into this one which had some CI specific files. but just reverted them, so should be good again.

roblatham00 commented on src/mpi/romio/adio/common/ad_write_str.c:
```diff
@@ -133,7 +133,7 @@ void ADIOI_GEN_WriteStrided(ADIO_File fd, const void *buf, int count,
     static char myname[] = "ADIOI_GEN_WriteStrided";
 
-    if (fd->hints->ds_write == ADIOI_HINT_DISABLE) {
+    if (1 || fd->hints->ds_write == ADIOI_HINT_DISABLE) {
```
uh... this shouldn't be here

mchaarawi replied:
I think the problem is that you are reviewing those patches one at a time.
can you just look at the github diff as if it's just 1 squashed commit?
On this page you can see just the changes:
https://github.com/pmodels/mpich/pull/3651/files

there are several intermediate patches or some accidental patches committed like this one that i reverted later.. i don't think you should be reviewing those, but just the final patch since you are going to squash it anyway, no?

roblatham00 commented:
there are 67 patches overall but I think I see some general feedback I can offer after looking at the first ten percent of them. 

I would like you to squash as many of these commits as possible.  You've got three years of development in here but we don't need that for the merge to MPICH master.  We need to a first order one patch: "Initial DAOS driver".  That patch will update common code where needed to teach it about daos -- configure.ac, ad_fs_type, ad_fs_proto, things like that.   

I don't know if you needed changes to common code, but if so, you might need a second patch (or perhaps I should say a 'zeroth' patch) if there a infrastructure things in ROMIO you fixed or want to make better.   For example (you might not have done any of these) maybe you did something to generalize hint processing, or you refactored the two phase collective code to avoid cutting and pasting a fourth version of ADIOI_Calc_aggregators (yes there are three of them.  I'm sorry!).  

Any tests you added to ROMIO or MPICH to exercise DAOS would be welcome, too.  Those would be a third patch.

roblatham00 commented:
sorry about repeating myself.  Can you then do the squashing?

mchaarawi commented:
ok here is a more simplified PR that should just have a single squashed commit.
i did manage to run the white space checker to actually fix all the issues, but i don't know why it's still failing in the PR. locally it passed the pre-commit hook.
https://github.com/pmodels/mpich/pull/4088

mchaarawi commented:
close as duplicate of https://github.com/pmodels/mpich/pull/4088
which squashes the history
